### ISSUE TITLE
Hit "Failed to send compaction finished" error when closing the DB.

### ISSUE THREAD

**Issue by vigneshc** [slatedb/slatedb#121]
With below code, I get this error. Likely compaction message is sent to invalid handles after close.

 `thread 'tokio-runtime-worker' panicked at /home/vigneshc/.cargo/git/checkouts/slatedb-beb4969f61fa9677/9642b33/src/compactor_executor.rs:130:18:
failed to send compaction finished msg: "SendError(..)"
`

```rust
let options = DbOptions {
        flush_ms: 100,
        manifest_poll_interval: Duration::from_millis(100),
        min_filter_keys: 10,
        l0_sst_size_bytes: 20 * 1024,
        compactor_options: Some(CompactorOptions::default()),
    };
    let kv_store = Db::open(
        Path::from("attempt3"),
        options,
        blob_store,
    )
    .await
    .unwrap();

    println!("Writing for next 1 minute");
    let mut record_count = 0;

    // write a loop that inserts 1000 keys repeatedly or a duration of 1 minute
    let start = Instant::now();
    let write_options = slatedb::config::WriteOptions {
        await_flush: false
    };

    while start.elapsed() < Duration::from_secs(60) {
        let iteration = start.elapsed().as_secs();
        for i in 0..1000 {

            kv_store.put_with_options(format!("key-{}-{}",iteration % 5, i).as_bytes(), format!("value{}", i).as_bytes(), &write_options).await;
            record_count += 1;
        }
        sleep(Duration::from_millis(100)).await;
    }

    println!("Total records written: {}", record_count);

    // Put
    let key = b"test_key";
    let value = b"test_value";
    kv_store.put(key, value).await;

    // Get
    println!("{:?}", kv_store.get(key).await.unwrap());

    // Delete
    kv_store.delete(key).await;
    println!("{:?}", kv_store.get(key).await.unwrap());

    // Close
    kv_store.close().await.unwrap();
```

**Comment by criccomini** [comment]
Tracing through this code a bit...

The error comes from TokioCompactionExecutorInner.start_compaction:

```rust
            this.worker_tx
                .send(CompactionFinished(result))
                .expect("failed to send compaction finished msg");
```

Called by TokioCompactionExecutor.start_compaction
Called by CompactorOrchestrator.start_compaction
Called by CompactorOrchestrator.submit_compaction
Called by CompactorOrchestrator.maybe_schedule_compactions
Called by CompactorOrchestrator.finish_compaction and CompactorOrchestrator.refresh_db_state
Called by CompactorOrchestrator.run (via `WorkerToOrchestoratorMsg::CompactionFinished` message in `loop`)
Sent by TokioCompactionExecutorInner.start_compaction

This looks like a loop, but it should halt when `maybe_schedule_compactions` returns false if there are < 4 L0 SSTs. In your case, I think the L0's are coming in fast enough (due to low `l0_sst_size_bytes`) that there are pretty much always >= 4 L0 SSTs.

As @vignesh suggested, I started looking at the close code. It's pretty simple:

The compactor's close happens in CompactorOrchestrator.run when it receives any message on `external_rx`:

```rust
    fn run(&mut self) {
        // ...
        loop {
            crossbeam_channel::select! {
                // ...
                recv(self.external_rx) -> _ => {
                    return;
                }
            }
        }
    }
```

`external_tx` and `external_rx` are created in Compactor.new
`external_tx` is saved as `main_tx` in the Compactor struct at the end of Compactor.new.
`self.main_tx.send(Shutdown).expect("main tx disconnected");` is sent on Compactor.close
Compactor.close is called by Db.close

So, it seems to me like it's indeed possible that close shuts down the CompactorOrchestrator run loop and destructs all resources (including `worker_tx`'s channel).

**Comment by criccomini** [comment]
I was able to replicate this bug as part of https://github.com/slatedb/slatedb/issues/124.

I see:

```
thread 'tokio-runtime-worker' panicked at /Users/chrisriccomini/Code/slatedb/src/db_common.rs:18:14:
failed to send memtable flush msg: SendError { .. }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at /Users/chrisriccomini/Code/slatedb/src/db.rs:323:30:
Failed to join flush thread: JoinError::Panic(Id(13), ...)
```

When I set the sleep duration in `drain_blocks` to `sleep(Duration::from_millis(40)).await;`. It's very hit and miss. I seem to be able to trigger it about once every 3-5 tries. I also get plenty of InvalidCompaction errors interspersed throughout the log (see #124).

_NOTE: I'm using the code in #124 to trigger these errors, not @vigneshc's code._

**Comment by criccomini** [comment]
Wondering: do we need a barrier on our loops that blocks prior to returning (when shutdown is called) until any outstanding flush/compaction calls are complete?

**Comment by rodesai** [comment]
> Also, there are three separate event loops: WAL flush (flush.rs), compaction (compactor.rs), and memtable flush (mem_table_flush.rs). I think I was the one who suggested breaking flush.rs up, but it's pretty hard to trace all this code. I think a single event loop might be preferable?

I had thought about this a bit. My reasoning for retaining the current breakup:
- I think flushing should be decoupled from memtable writes because we don't want to block flushes on memtable writes because writes block on the flush task.
- Compaction is nice to keep decoupled from memtable/write flushing because it should be runnable as a separate process. That said, we could in theory opt to run it in the same event loop as memtable flush if compaction is running in the db (the common case for now). The main hurdle here is that right now the compaction event loop is a thread, and the memtable flush is an async task. If we want to merge them I'd prefer making the memtable flush non-async and running it in a thread that also runs the compactor.
- All that said, I wasn't convinced of the benefit. At the end of the day these are 3 distinct processes that aren't inherently coupled to each other - seems easiest to reason about if they are modeled as separate tasks/threads.

> Wondering: do we need a barrier on our loops that blocks prior to returning (when shutdown is called) until any outstanding flush/compaction calls are complete?

Did you have a use case in mind? We won't do anything "wrong" if we don't do this. The main benefit I can see is that we can do a clean shutdown where we flush whatever we have in the memtable to an l0 so we don't have to recover it. But we could also acheive this by shutting down the flush task first, and then shutting down the memtable task.

**Comment by rodesai** [comment]
For this bug, we need to implement graceful shutdown of the compaction executor. As @criccomini observed above, right now we just drop the executor on the floor when `CompactorOrchestrator` stops, without trying to stop in-flight compaction tasks. This won't break anything, but it can lead to the spurious error above, and more egregiously - leave the compaction task hanging around chewing up resources if closing the db isn't done right before shutting down tokio.

I think the shutdown flow should look something like:

- Compactor Orchestrator is notified that it should stop over `external_rx`
- Compactor Orchestrator asks the compaction executor to stop (using some function like `stop` in `CompactionExecutor`).
- The `CompactionExecutor` impl implements `stop` by notifying any compaction tasks that they should stop, and then joining outstanding tasks
- Compaction tasks need to check whether they need to stop, and should exit when they do.
- Compactor Orchestrator should drain the worker status channel (`worker_rx`) and make any updates to the manifest about completed compaction work.
- Compactor Orchestrator exits.

**Comment by criccomini** [comment]
> But we could also acheive this by shutting down the flush task first, and then shutting down the memtable task.

Gotcha. Yes, I misunderstood select! and thought it could run branches in parallel. Sounds like it's [only concurrent](https://docs.rs/tokio/latest/tokio/macro.select.html#runtime-characteristics). Thus, once flush's shutdown message is received, no further flushes can happen. So no barrier is needed.

**Comment by criccomini** [comment]
Taking a shot at this.

**Comment by criccomini** [comment]
PR up at #133
